theano/gof/type.py

```diff
@@ -795,3 +796,137 @@ def signature(self):
         return (self.type,)
 
 CDataType.Constant = CDataTypeConstant
+
+
+class EnumType(Type, dict):
```

Comment:
The problem this is facing is that we want enum values in C and some sort of value in python to indicate what we want.

We can't just use the C values in python since we don't know what they are.  We can't really use the define name as a python value because this will be super painful to interact with from the C side, so the compromise is this.  A type that will define a bunch of values in python and C and can translate between each other.

Maybe we could have something that uses ints or whatever in python and the actual define values in C to have the best of both worlds.  In this case the python values wouldn't have any relation to the actual C values, but the type itself would have a mapping and use it to translate.  However I haven't done the work myself, so I'm not sure if it makes sense.